
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Xinmeng Liao">
      
      
        <link rel="canonical" href="https://xinmengliao.github.io/PhenomePortal-VarXOmics/Functions/Small%20Variant%20Interpretation/">
      
      
        <link rel="prev" href="../CNV%20and%20SV%20Query/">
      
      
        <link rel="next" href="../Gene%20Query/">
      
      
      <link rel="icon" href="../../images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.22">
    
    
      
        <title>Small Variant Interpretation - VarXOmics Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.84d31ad4.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#multi-variant-analysis" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="VarXOmics Documentation" class="md-header__button md-logo" aria-label="VarXOmics Documentation" data-md-component="logo">
      
  <img src="../../images/white-logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            VarXOmics Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Small Variant Interpretation
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="blue"  aria-hidden="true"  type="radio" name="__palette" id="__palette_0">
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/XinmengLiao/PhenomePortal-VarXOmics" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    XinmengLiao/PhenomePortal-VarXOmics
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../Intro/" class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../Small%20Variant%20Query/" class="md-tabs__link">
          
  
  
  Tutorials

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../Data%20Access/" class="md-tabs__link">
        
  
  
    
  
  Data Access

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="VarXOmics Documentation" class="md-nav__button md-logo" aria-label="VarXOmics Documentation" data-md-component="logo">
      
  <img src="../../images/white-logo.png" alt="logo">

    </a>
    VarXOmics Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/XinmengLiao/PhenomePortal-VarXOmics" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    XinmengLiao/PhenomePortal-VarXOmics
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Small%20Variant%20Query/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Small Variant Query
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../CNV%20and%20SV%20Query/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Structural Variant Query
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Small Variant Interpretation
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Small Variant Interpretation
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#multi-variant-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      🔹 Multi-Variant Analysis
    </span>
  </a>
  
    <nav class="md-nav" aria-label="🔹 Multi-Variant Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-input-data" class="md-nav__link">
    <span class="md-ellipsis">
      1. Input Data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-variant-filtering-settings" class="md-nav__link">
    <span class="md-ellipsis">
      2. Variant Filtering Settings
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sample-filters" class="md-nav__link">
    <span class="md-ellipsis">
      🧩 Sample Filters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#allele-frequency-filters" class="md-nav__link">
    <span class="md-ellipsis">
      🌍 Allele Frequency Filters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#predictive-scores" class="md-nav__link">
    <span class="md-ellipsis">
      🤖 Predictive Scores
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#spliceai-thresholds" class="md-nav__link">
    <span class="md-ellipsis">
      🧬 SpliceAI Thresholds
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bayesdel-scores" class="md-nav__link">
    <span class="md-ellipsis">
      🧮 BayesDel Scores
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#classification-filters" class="md-nav__link">
    <span class="md-ellipsis">
      🧾 Classification Filters
    </span>
  </a>
  
    <nav class="md-nav" aria-label="🧾 Classification Filters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#alphamissense-am-classification" class="md-nav__link">
    <span class="md-ellipsis">
      AlphaMissense (AM) Classification
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#alphamissense-am-pathogenicity" class="md-nav__link">
    <span class="md-ellipsis">
      AlphaMissense (AM) Pathogenicity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clinvar-status" class="md-nav__link">
    <span class="md-ellipsis">
      ClinVar Status
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#acmg-classification" class="md-nav__link">
    <span class="md-ellipsis">
      ACMG Classification
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-run-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      3. Run Analysis
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-result-overview" class="md-nav__link">
    <span class="md-ellipsis">
      4. Result Overview
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. Result Overview">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#variant-overview" class="md-nav__link">
    <span class="md-ellipsis">
      🧬 Variant Overview
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variant-summary" class="md-nav__link">
    <span class="md-ellipsis">
      📊 Variant Summary
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#enrichment-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      🔬 Enrichment Analysis
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gwas-qtl-associations" class="md-nav__link">
    <span class="md-ellipsis">
      📚 GWAS &amp; QTL Associations
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#network" class="md-nav__link">
    <span class="md-ellipsis">
      🕸️ Network
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#exomiser" class="md-nav__link">
    <span class="md-ellipsis">
      🧠 Exomiser
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-export-results" class="md-nav__link">
    <span class="md-ellipsis">
      5. Export results
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Gene%20Query/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Gene Query
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../CNVs%20and%20SVs%20intrepretation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    CNVs and SVs Interpretation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Data%20Access/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data Access
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>Small Variant Interpretation</h1>

<h2 id="multi-variant-analysis">🔹 Multi-Variant Analysis<a class="headerlink" href="#multi-variant-analysis" title="Permanent link">&para;</a></h2>
<p>VarXOmics supports <strong>multi-variant analysis</strong> for <strong>small variants (SNPs &amp; INDELs)</strong>, <strong>copy number variants (CNVs)</strong>, and <strong>structural variants (SVs)</strong>.<br />
This module allows users to upload VCF files, configure advanced filtering parameters, and perform variant interpretation in a structured and reproducible manner.</p>
<h3 id="1-input-data">1. Input Data<a class="headerlink" href="#1-input-data" title="Permanent link">&para;</a></h3>
<ul>
<li>From the <strong>Home</strong> page, choose <strong>“Multi Variant”</strong>.  </li>
<li>Drop and upload a <strong>VCF file</strong>.  </li>
<li>Click <strong>Run</strong> to proceed to the variant filtering configuration window.</li>
</ul>
<p><img width="60%" src="https://github.com/user-attachments/assets/9c37f2db-ab7e-4c04-951d-db788373fcef" /></p>
<h3 id="2-variant-filtering-settings">2. Variant Filtering Settings<a class="headerlink" href="#2-variant-filtering-settings" title="Permanent link">&para;</a></h3>
<p>After uploading the VCF, a configuration panel titled <strong>“Multi-Variant Analysis Settings”</strong> will appear.<br />
This panel allows fine-tuning of variant filtration and prioritization parameters.</p>
<h3 id="sample-filters">🧩 Sample Filters<a class="headerlink" href="#sample-filters" title="Permanent link">&para;</a></h3>
<table>
<thead>
<tr>
<th>Setting</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Reference Genome</strong></td>
<td>Select the genome build used for alignment — <strong>GRCh37</strong> or <strong>GRCh38</strong>.</td>
</tr>
<tr>
<td><strong>Gender</strong></td>
<td>Choose the biological sex of the sample (<strong>Male</strong>, <strong>Female</strong>, or <strong>Unknown</strong>) for filtering X/Y-linked variants.</td>
</tr>
<tr>
<td><strong>Only PASS Variants</strong></td>
<td>When set to <strong>Yes</strong>, filters out non-PASS variants from the VCF.</td>
</tr>
<tr>
<td><strong>Inheritance Pattern</strong></td>
<td>Specify the expected mode of inheritance (<strong>Dominant</strong>, <strong>Recessive</strong>, <strong>Compound Heterozygous</strong>, or <strong>Any pattern</strong>).</td>
</tr>
<tr>
<td><strong>HPO Terms (CSV)</strong></td>
<td>Provide one or multiple <strong>HPO IDs</strong> (comma-separated) to integrate phenotype-based filtering, e.g. <code>HP:0001250,HP:0000707</code>.</td>
</tr>
</tbody>
</table>
<h3 id="allele-frequency-filters">🌍 Allele Frequency Filters<a class="headerlink" href="#allele-frequency-filters" title="Permanent link">&para;</a></h3>
<table>
<thead>
<tr>
<th>Setting</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>AF ClinVar</strong></td>
<td>Sets a frequency cap for ClinVar-derived evidence. Variants above this threshold are excluded.</td>
</tr>
<tr>
<td><strong>AF (Predicted variants)</strong></td>
<td>Defines the <strong>allele frequency cutoff</strong> for predicted variants (e.g., 0.05). Lower thresholds retain rarer variants.</td>
</tr>
</tbody>
</table>
<h3 id="predictive-scores">🤖 Predictive Scores<a class="headerlink" href="#predictive-scores" title="Permanent link">&para;</a></h3>
<table>
<thead>
<tr>
<th>Score</th>
<th>Description</th>
<th>Typical Cutoff</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>ADA</strong></td>
<td>ADA score for splicing impact prediction.</td>
<td>0.6</td>
</tr>
<tr>
<td><strong>RF</strong></td>
<td>Random Forest model score for splicing variant prediction.</td>
<td>0.6</td>
</tr>
<tr>
<td><strong>REVEL</strong></td>
<td>REVEL score for missense variant pathogenicity prediction.</td>
<td>0.75</td>
</tr>
</tbody>
</table>
<h3 id="spliceai-thresholds">🧬 SpliceAI Thresholds<a class="headerlink" href="#spliceai-thresholds" title="Permanent link">&para;</a></h3>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Acceptor Loss (AL)</strong></td>
<td>Minimum threshold for acceptor site loss prediction.</td>
<td>0.5</td>
</tr>
<tr>
<td><strong>Acceptor Gain (AG)</strong></td>
<td>Minimum threshold for acceptor site gain prediction.</td>
<td>0.5</td>
</tr>
<tr>
<td><strong>Donor Loss (DL)</strong></td>
<td>Minimum threshold for donor site loss prediction.</td>
<td>0.5</td>
</tr>
<tr>
<td><strong>Donor Gain (DG)</strong></td>
<td>Minimum threshold for donor site gain prediction.</td>
<td>0.5</td>
</tr>
</tbody>
</table>
<h3 id="bayesdel-scores">🧮 BayesDel Scores<a class="headerlink" href="#bayesdel-scores" title="Permanent link">&para;</a></h3>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>BayesDel (with AF)</strong></td>
<td>Bayesian deleteriousness score integrating allele frequency.</td>
<td>0.069</td>
</tr>
<tr>
<td><strong>BayesDel (no AF)</strong></td>
<td>Bayesian deleteriousness score excluding frequency adjustment.</td>
<td>-0.057</td>
</tr>
</tbody>
</table>
<blockquote>
<p>These thresholds align with backend default values for consistent filtering.</p>
</blockquote>
<h3 id="classification-filters">🧾 Classification Filters<a class="headerlink" href="#classification-filters" title="Permanent link">&para;</a></h3>
<h4 id="alphamissense-am-classification"><strong>AlphaMissense (AM) Classification</strong><a class="headerlink" href="#alphamissense-am-classification" title="Permanent link">&para;</a></h4>
<p>Select which <strong>ACMG/AMP pathogenicity tiers</strong> to retain in the analysis:
- Pathogenic<br />
- Likely Pathogenic<br />
- Ambiguous<br />
- Benign  </p>
<h4 id="alphamissense-am-pathogenicity"><strong>AlphaMissense (AM) Pathogenicity</strong><a class="headerlink" href="#alphamissense-am-pathogenicity" title="Permanent link">&para;</a></h4>
<ul>
<li>Numerical cutoff representing aggregated pathogenicity evidence (e.g., 0.564).</li>
</ul>
<h4 id="clinvar-status"><strong>ClinVar Status</strong><a class="headerlink" href="#clinvar-status" title="Permanent link">&para;</a></h4>
<p>Select ClinVar review statuses to include:
- Pathogenic<br />
- Likely Pathogenic<br />
- Uncertain Significance<br />
- Conflicting Classifications<br />
- Benign<br />
- Likely Benign  </p>
<h4 id="acmg-classification"><strong>ACMG Classification</strong><a class="headerlink" href="#acmg-classification" title="Permanent link">&para;</a></h4>
<p>Specify ACMG evidence tiers for filtering:
- Pathogenic<br />
- Likely Pathogenic<br />
- Uncertain Significance<br />
- Benign<br />
- Likely Benign  </p>
<h3 id="3-run-analysis">3. Run Analysis<a class="headerlink" href="#3-run-analysis" title="Permanent link">&para;</a></h3>
<p>Once all filters are configured, click <strong>Start Analysis</strong> to execute the variant prioritization pipeline.  </p>
<h3 id="4-result-overview">4. Result Overview<a class="headerlink" href="#4-result-overview" title="Permanent link">&para;</a></h3>
<p>VarXOmics processes and prioritizes all variants based on its integrated filtering and scoring strategy. The final results are divided into six major sections, each providing an in-depth view of the analysis outcome.</p>
<hr />
<h4 id="variant-overview">🧬 Variant Overview<a class="headerlink" href="#variant-overview" title="Permanent link">&para;</a></h4>
<p>This section displays all filtered variants prioritized by the VarXOmics strategy. Each variant is annotated comprehensively, including genomic, transcript, clinical, and functional evidence.</p>
<p><strong>Key Features:</strong>
- <strong>Detailed Annotation Table:</strong><br />
  Displays complete variant-level information.<br />
  Users can customize visible columns using the <strong>“Columns”</strong> button.
- <strong>Quick Summary Subpanels:</strong><br />
  Provides concise visual summaries of:
  - <strong>Variant Consequence Distribution</strong><br />
  - <strong>Variant Types</strong><br />
  - <strong>Clinical Pathogenicity</strong><br />
  - <strong>Affected Genes</strong></p>
<p>These visual overviews allow users to rapidly assess the biological and clinical relevance of the prioritized variants.  </p>
<p><img width="90%" src="https://github.com/user-attachments/assets/0e3f9b39-147b-4f3f-901b-56a42b6fe1ff" /></p>
<hr />
<h4 id="variant-summary">📊 Variant Summary<a class="headerlink" href="#variant-summary" title="Permanent link">&para;</a></h4>
<p>This section provides an integrated summary of all filtered variants through seven visual plots, allowing users to explore the overall variant landscape.</p>
<p><strong>Displayed Figures:</strong>
  - <strong>SNP Density Plot:</strong><br />
   Visualizes the distribution of SNPs across all chromosomes.
  - <strong>ClinVar Ideogram:</strong><br />
   Shows the chromosomal positions of ClinVar-annotated variants and their pathogenicity status.
  - <strong>ClinVar Pathogenicity:</strong><br />
   Summarizes the proportion of variants categorized by ClinVar pathogenicity.
  - <strong>Variant Class:</strong><br />
   Displays the proportion of SNPs, INDELs, and other variant types.
  - <strong>Top 10 Disease Accumulators:</strong><br />
   Highlights the diseases most frequently associated with the filtered variants.
  - <strong>Top 10 Gene Accumulators:</strong><br />
   Shows the genes where variants are most enriched.
  - <strong>Variant Consequence Bar Plot:</strong><br />
   Displays the frequency of different variant consequences (e.g., missense, nonsense, synonymous).</p>
<p>Together, these plots offer an intuitive view of the biological and clinical impact of the variant set.</p>
<hr />
<h4 id="enrichment-analysis">🔬 Enrichment Analysis<a class="headerlink" href="#enrichment-analysis" title="Permanent link">&para;</a></h4>
<p>This section performs functional enrichment for all genes linked to the filtered variants, helping users interpret the biological processes and pathways involved.</p>
<p><strong>Components:</strong>
- <strong>GO Term Enrichment Table:</strong><br />
  Lists significantly enriched <strong>Gene Ontology (GO)</strong> terms (Biological Process, Molecular Function, Cellular Component).
- <strong>KEGG Pathway Table:</strong><br />
  Shows <strong>KEGG pathway enrichment</strong> results for the filtered gene set.
- <strong>GO Enrichment Visualization:</strong><br />
  - Top enriched GO terms ranked by gene count.
  - GO Term Counts.
  - Statistics of GO Biological Processes, Molecular Functions, Cellular Components. 
- <strong>KEGG Enrichment Visualization:</strong><br />
  Presents top pathways ranked by enrichment significance.</p>
<p>The enrichment section provides direct insight into the biological themes underlying the filtered variant set.</p>
<p><img width="100%" src="https://github.com/user-attachments/assets/6971a902-0810-4f0d-8c8b-0dc60ae77bfc" /></p>
<hr />
<h4 id="gwas-qtl-associations">📚 GWAS &amp; QTL Associations<a class="headerlink" href="#gwas-qtl-associations" title="Permanent link">&para;</a></h4>
<p>This section presents variant- and gene-level associations collected from multiple public resources, allowing users to explore known evidence linking the queried variants or genes to complex traits, diseases, and regulatory effects.<br />
Contents cover: GWAS, eQTL, pQTL, Pharmacogenomics (PGx).<br />
Users can customize visible columns with the <strong>“Columns”</strong> button (top-right corner).<br />
Together, these resources allow users to assess the translational and regulatory significance of their filtered variants.</p>
<hr />
<h4 id="network">🕸️  Network<a class="headerlink" href="#network" title="Permanent link">&para;</a></h4>
<p>This section visualizes all biological interactions derived from the filtered variant and gene set.<br />
This integrative network provides a global view of how variants, genes, proteins, drugs, and phenotypes interconnect through multiple evidence layers.<br />
This network provides an integrated, evidence-weighted visualization of how variants influence genes, pathways, and drug responses — bridging molecular biology and clinical relevance.</p>
<p><strong>Interactive Features:</strong>
- <strong>Edge Weight Filter:</strong><br />
  Adjust the slider to dynamically filter edges weights. For each interactioins, the weights are set as:</p>
<table>
<thead>
<tr>
<th>Interaction Type</th>
<th>Weight / Significance Cutoff</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>PPIs</strong></td>
<td>Interaction confidence <strong>&gt; 0.7</strong></td>
</tr>
<tr>
<td><strong>GWAS</strong></td>
<td><strong>p &lt; 5 × 10⁻⁸</strong></td>
</tr>
<tr>
<td><strong>eQTL / pQTL</strong></td>
<td><strong>p_beta &lt; 0.05</strong></td>
</tr>
<tr>
<td><strong>MR (Mendelian Randomization)</strong></td>
<td><strong>p &lt; 0.05</strong></td>
</tr>
</tbody>
</table>
<ul>
<li><strong>Search Box:</strong><br />
  Locate nodes by label (e.g., gene symbol, variant ID, drug name).</li>
<li><strong>Node Summary Panel:</strong><br />
  Displays total counts of nodes and edges, categorized by node and interaction type.</li>
<li><strong>Table View:</strong><br />
  Click the <strong>table icon</strong> in the bottom-right corner to open a structured view of the network, allowing users to inspect and filter interaction data directly.</li>
</ul>
<p>✅ <strong>Tip:</strong><br />
Users can combine interaction filters and weight thresholds to create <strong>focused sub-networks</strong>, such as <em>gene–drug pharmacogenomic networks</em> or <em>variant–phenotype associations</em> for specific pathways.</p>
<p><img width="100%" src="https://github.com/user-attachments/assets/fd1a0e3d-1ad4-4de6-93fb-b363d3b144d0" /></p>
<hr />
<h4 id="exomiser">🧠 Exomiser<a class="headerlink" href="#exomiser" title="Permanent link">&para;</a></h4>
<p>VarXOmics integrates <a href="https://exomiser.readthedocs.io/en/stable/#"><strong>Exomiser</strong></a> analysis to support phenotype-driven variant prioritization. By linking variant annotations with <strong>HPO terms</strong>, Exomiser scores potential candidate variants and genes according to their phenotypic relevance.</p>
<p>This section presents:<br />
- Ranked list of candidate genes and variants associated with input HPO phenotypes.<br />
- Exomiser scores indicating phenotype similarity and variant pathogenicity.<br />
- Cross-references with ClinVar and other annotations for interpretation consistency.</p>
<p><img width="100%" src="https://github.com/user-attachments/assets/df783ae6-6132-4637-997d-2036a552fe3c" /></p>
<hr />
<h3 id="5-export-results">5. Export results<a class="headerlink" href="#5-export-results" title="Permanent link">&para;</a></h3>
<p>Users can export all results, figures, and enrichment tables as a single compressed archive by clicking <strong>“Export All”</strong> in the top-right corner of the interface.</p>







  
    
  
  


  <aside class="md-source-file">
    
      
  <span class="md-source-file__fact">
    <span class="md-icon" title="Last update">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg>
    </span>
    <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date" title="October 30, 2025 12:22:15 UTC">October 30, 2025</span>
  </span>

    
    
    
    
  </aside>





                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      © 2025 VarXOmics Team
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.sections", "toc.integrate", "content.code.copy"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
    
  </body>
</html>