
## 🔹 CNV & SV Query

### 1. Input Variant  
- From the **Home** page, choose **“Single CNV/SV”**.  
- Enter the variant in the format: `chromosome_starting position_ending position_mutation type`
  Example: `chr3_100000_200000_dup`  
- Click **Run** to proceed.  

  <img width="60%" src="https://github.com/user-attachments/assets/bbd74e94-502e-4ae4-908f-3364e0103a66" />

<br>

### 2. Select Reference Genome and Type  
- Reference genome options: **GRCh37**, **GRCh38**  
- Mutation type options: **CNV**, **SV**  
- Click **Start Analysis** to begin.  
> **Note:** CNVs and SVs are annotated using **AnnotSV**.  

   <img width="65%" src="https://github.com/user-attachments/assets/c2cbda85-2c06-48d9-8e68-109e58665289" />

<br>

### 3. View Results  
- Annotated results appear once loaded.  
- Customize displayed columns using the **“Columns”** button.

### 4. Export Data  
- Export all annotation tables as a **compressed file** via the **“Export All”** button.
<img width="1089" height="213" alt="image" src="https://github.com/user-attachments/assets/1e9f08cc-5c5e-4438-b7f9-9351f57d3549" />
