---
title: "MultiVariant Figure"
author: "Xinmeng Liao"
date: "2025-06-13"
output: html_document
---

```{r setup, include=FALSE}
library(dplyr)
library(tidyverse)
library(data.table)
library(clusterProfiler)
library(org.Hs.eg.db)
library(ggsci)
library(viridis)
for (i in c("select","filter", "mutate","rename", "left_join", "slice")){
  conflicted::conflict_prefer(i, "dplyr")
}
rm(i)
conflicted::conflicts_prefer(stats::sd)
conflicted::conflicts_prefer(httr::content)
conflicted::conflicts_prefer(plotly::layout)
```

# 1. Load files 
```{r}
result_file <- "/Users/xinmengliao/Documents/Project/20250516_Webserver/Case_Study/Healthy/WBWG_01_P001_30.nodup.txt"
result <- read.csv(result_file,header = T,sep = "\t")

colnames(result) <- gsub("am_","AlphaMissense_", colnames(result))
result <- result %>% 
  mutate(rsID = sapply(strsplit(split = "&", Existing_variation),`[`,1)) %>% 
  mutate(rsID = if_else(grepl("^rs", rsID,fixed = F), rsID, NA)) %>% unique() 
output_file <- "/Users/xinmengliao/Documents/Project/20250516_Webserver/usr_output/Variant/MultiVariant/multivariant_output.txt"
```


# Variant class (Pie chart)
```{r}
variant.class <- result %>%
        select(variant_info, VARIANT_CLASS) %>% unique() %>% 
        mutate(VARIANT_CLASS = case_when(VARIANT_CLASS == "insertion" ~ "INS", 
                                         VARIANT_CLASS == "deletion" ~ "DEL", TRUE ~ VARIANT_CLASS)) %>% 
        group_by(VARIANT_CLASS) %>% 
        summarise(count = n(), .groups = "drop") %>% 
        arrange(desc(count)) %>% 
        mutate(percent = round(count / sum(count) * 100, 1),
               label = paste0(VARIANT_CLASS,  " (", percent, "%)")) %>% 
        mutate(color = case_when(VARIANT_CLASS == "SNV" ~ "#4575b4",
                                 VARIANT_CLASS == "INS" ~ "#fc8d59",
                                 VARIANT_CLASS == "DEL" ~ "#1a9850", TRUE ~ "#999999"))
write.table(variant.class, gsub("txt","varint_class_pie.txt", output_file),quote = F,sep = "\t",row.names = F)

variant.class.p <- ggplot(variant.class, aes(x = "", y = count, fill = VARIANT_CLASS)) +
  geom_bar(stat = "identity", width = 1, color = "white") +
  coord_polar(theta = "y") +
  geom_text(aes(label = label), 
            position = position_stack(vjust = 0.5), 
            color = "white", size = 4) +
  scale_fill_manual(values = c("SNV" = "#4575b4","DEL" = "#1a9850", "INS" = "#fc8d59")) +
  theme_void() +
  theme(legend.position = "none")

ggsave(variant.class.p, filename = gsub("txt","varint_class_pie.png", output_file),dpi = 300)
```

# Consequence (bar plot)
```{r}
variant.conseq <- result %>%
        select(variant_info, Consequence) %>%
        unique() %>%
        mutate(Consequence = sapply(strsplit(split = "&",Consequence),`[`,1)) %>% 
        group_by(Consequence) %>%
        summarise(count = n(), .groups = "drop") %>%
        arrange(desc(count)) 
write.table(variant.conseq, gsub("txt","varint_conseq_bar.txt", output_file),quote = F,sep = "\t",row.names = F)

fill_levels <- n_distinct(variant.conseq$count)
color_scale <- if (fill_levels > 10) {
        scale_fill_viridis(option = "A", discrete = TRUE)
        } else {
          ggsci::scale_fill_npg()
        }
variant.conseq$fill_value <- if (fill_levels > 10) variant.conseq$count else as.factor(variant.conseq$count)
variant.conseq.p <- ggplot(variant.conseq, aes(x = reorder(Consequence, count), y = count, fill = Consequence)) +
        geom_bar(stat = "identity") +
        color_scale +
        labs(x = "", y = "No. of Variants") +
        theme_minimal() +
        theme(
          legend.position = "none",
          panel.grid.major.y = element_blank(),
          panel.grid.minor.y = element_blank(),
          panel.grid.major.x = element_line(color = "#f0f0f0"),
          panel.grid.minor.x = element_blank(),
          axis.text.y = element_text(size = 11),
          axis.text.x = element_text(size = 10,angle = 270, hjust = 0),
          axis.title.x = element_text(size = 12, margin = margin(t = 10))
        )

ggsave(variant.conseq.p, filename = gsub("txt","varint_conseq_bar.png", output_file),dpi = 300)
```

# Allele Frequency 
```{r}
variant.af <- result %>% select(variant_info, MAX_AF) %>%  unique() %>% 
  mutate(MAX_AF = if_else(is.na(MAX_AF), 0 , MAX_AF)) %>% unique()
write.table(variant.af, gsub("txt","varint_af_density.txt", output_file),quote = F,sep = "\t",row.names = F)

variant.af.p <- ggplot(variant.af, aes(x = MAX_AF)) +
  geom_density(fill = "mediumturquoise", alpha = 0.8) +
  scale_x_continuous(breaks = seq(0, 1, 0.15), limits = c(0, 1)) +
  labs(
    x = "Maximum Global Allele Frequency",
    y = NULL
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(),
    axis.title.y = element_blank(),
    plot.title = element_text(hjust = 0, face = "bold"),
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    plot.background = element_rect(fill = "#f0f0f0", color = NA),
    panel.background = element_rect(fill = "#f0f0f0", color = NA)
  )

ggsave(variant.af.p, filename = gsub("txt","varint_af_density.png", output_file),dpi = 300)
```

# Gene burden (bar plot) 
```{r}
gene.burden <- result %>% 
        select(variant_info, SYMBOL) %>% unique() %>% 
        group_by(SYMBOL) %>% 
        summarise(count = n(), .groups = "drop") %>% 
        arrange(desc(count),SYMBOL) 

fill_levels <- n_distinct(gene.burden$count)
color_scale <- if (fill_levels > 10) {
        scale_fill_viridis(option = "A", discrete = TRUE)
        } else {
          ggsci::scale_fill_npg()
        }
gene.burden$fill_value <- if (fill_levels > 10) gene.burden$count else as.factor(gene.burden$count)
write.table(gene.burden, gsub("txt","gene_burden_bar.txt", output_file),quote = F,sep = "\t",row.names = F)


if(nrow(gene.burden) > 10){
  gene.burden_sub <- gene.burden %>% slice(1:10)
}else{
  gene.burden_sub <- gene.burden
}

gene.burden.p <- ggplot(gene.burden, aes(y = reorder(SYMBOL, count), x = count, fill = SYMBOL)) +
  geom_bar(stat = "identity") +
  color_scale +
  labs(y = "", x = "Number of Variants") +
  theme_minimal() +
  theme(
    legend.position = "none",
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    panel.grid.major.x = element_line(color = "#f0f0f0"),
    panel.grid.minor.x = element_blank(),
    axis.text.y = element_text(size = 11),
    axis.text.x = element_text(size = 10),
    axis.title.x = element_text(size = 12, margin = margin(t = 10))
  )+
  scale_x_continuous(breaks =  seq(min(gene.burden$count), max(gene.burden$count),1))

ggsave(gene.burden.p, filename = gsub("txt","gene_burden_bar.png", output_file),dpi = 300)
```


# Disease burden (bar plot)
```{r}
disease.burden <- result %>% 
        select(variant_info, ClinVar_CLNDN) %>% unique() %>%
  filter(ClinVar_CLNDN != "") %>% 
  separate_rows(ClinVar_CLNDN, sep = "&(?!_)") %>% 
  filter(ClinVar_CLNDN != "not_provided" & ClinVar_CLNDN != "not_specified") %>% 
  group_by(ClinVar_CLNDN) %>% 
  summarise(count = n(), .groups = "drop") %>% 
  arrange(desc(count),ClinVar_CLNDN) 

fill_levels <- n_distinct(disease.burden$count)
color_scale <- if (fill_levels > 10) {
        scale_fill_viridis(option = "A", discrete = TRUE)
        } else {
          ggsci::scale_fill_npg()
        }
disease.burden$fill_value <- if (fill_levels > 10) disease.burden$count else as.factor(disease.burden$count)
write.table(disease.burden, gsub("txt","disease_burden_bar.txt", output_file),quote = F,sep = "\t",row.names = F)

if(nrow(disease.burden) > 10){
  disease.burden_sub <- disease.burden %>% slice(1:10)
}else{
  disease.burden_sub <- disease.burden
}

disease.burden.p <- ggplot(disease.burden_sub, aes(y = reorder(ClinVar_CLNDN, count), x = count, fill = ClinVar_CLNDN)) +
  geom_bar(stat = "identity") +
  color_scale +
  labs(y = "", x = "Number of Variants") +
  theme_minimal() +
  theme(
    legend.position = "none",
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    panel.grid.major.x = element_line(color = "#f0f0f0"),
    panel.grid.minor.x = element_blank(),
    axis.text.y = element_text(size = 11),
    axis.text.x = element_text(size = 10),
    axis.title.x = element_text(size = 12, margin = margin(t = 10))
  )+
  scale_x_continuous(breaks =  seq(min(disease.burden$count), max(disease.burden$count),1))

ggsave(disease.burden.p, filename = gsub("txt","disease_burden_bar.png", output_file),dpi = 300)
```


# Erichment analysis 
```{r}
gene_list <- result$SYMBOL %>% unique()
gene_df <- bitr(gene_list, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Hs.eg.db)

# GO-BP
if (nrow(gene_df) > 0) {
  enrich_res_gocc <- enrichGO(gene = gene_df$ENTREZID,
                              OrgDb = org.Hs.eg.db,
                              ont = "CC",
                              pAdjustMethod = "BH",
                              pvalueCutoff = 1,
                              readable = TRUE) %>% 
    as.data.frame() %>% mutate(Source  = " GO-CC")
}else{
  enrich_res_gocc <- data.frame()
}

if(nrow(enrich_res_gocc) > 0 ){
  if(nrow(enrich_res_gocc)>= 10 ){
  gocc_plot <- enrich_res_gocc %>% arrange(p.adjust) %>% slice(1:10)
  }else if(nrow(enrich_res_gocc) < 10 ){
    gocc_plot <- enrich_res_gocc %>% arrange(p.adjust)
  }
  
  gocc_plot$Description <- gsub("_", " ", gocc_plot$Description)
  gocc_plot$Description <- factor(gocc_plot$Description, levels = rev(gocc_plot$Description))
  gocc_plot$significance <- -log10(gocc_plot$p.adjust)
  res_num <- nrow(gocc_plot)
  
  gocc_plot.p <- ggplot(gocc_plot, aes(x = significance, y = Description, color = round(p.adjust, 2))) +
      geom_point(size = 3) +
      scale_color_gradient(low = "red", high = "blue") +
      labs(x = "-log10(p.adj)", y = "", color = "p.adj", size = "Gene count", title = paste("Top ", res_num, " enriched GO-CC terms")) +
      theme_classic() +
      theme(
        axis.text.y = element_text(size = 11),
        legend.position = "right")
  ggsave(gocc_plot.p, filename = gsub("txt","gocc_plot.png",output_file),dpi = 300)
}
  

# GO-BP
if (nrow(gene_df) > 0) {
  enrich_res_gobp <- enrichGO(gene = gene_df$ENTREZID,
                              OrgDb = org.Hs.eg.db,
                              ont = "BP",
                              pAdjustMethod = "BH",
                              pvalueCutoff = 1,
                              readable = TRUE) %>% 
    as.data.frame() %>% mutate(Source  = " GO-BP")
}else{
  enrich_res_gobp <- data.frame()
}

if(nrow(enrich_res_gobp) > 0 ){
  if(nrow(enrich_res_gobp)>= 10 ){
  gobp_plot <- enrich_res_gobp %>% arrange(p.adjust) %>% slice(1:10)
  }else if(nrow(enrich_res_gobp) < 10 ){
    gobp_plot <- enrich_res_gobp %>% arrange(p.adjust)
  }
  
  gobp_plot$Description <- gsub("_", " ", gobp_plot$Description)
  gobp_plot$Description <- factor(gobp_plot$Description, levels = rev(gobp_plot$Description))
  gobp_plot$significance <- -log10(gobp_plot$p.adjust)
  res_num <- nrow(gobp_plot)
  
  gobp_plot.p <- ggplot(gobp_plot, aes(x = significance, y = Description, color = round(p.adjust, 2))) +
      geom_point(size = 3) +
      scale_color_gradient(low = "red", high = "blue") +
      labs(x = "-log10(p.adj)", y = "", color = "p.adj", size = "Gene count", title = paste("Top ", res_num, " enriched GO-BP terms")) +
      theme_classic() +
      theme(
        axis.text.y = element_text(size = 11),
        legend.position = "right")
  ggsave(gobp_plot.p, filename = gsub("txt","gobp_plot.png",output_file),dpi = 300)
}
  
# GO-MF
if (nrow(gene_df) > 0) {
  enrich_res_gomf <- enrichGO(gene = gene_df$ENTREZID,
                              OrgDb = org.Hs.eg.db,
                              ont = "MF",
                              pAdjustMethod = "BH",
                              pvalueCutoff = 1,
                              readable = TRUE) %>% 
    as.data.frame() %>% mutate(Source  = " GO-MF")
}else{
  enrich_res_gomf <- data.frame()
}

if(nrow(enrich_res_gomf) > 0 ){
  if(nrow(enrich_res_gomf)>= 10 ){
  gomf_plot <- enrich_res_gomf %>% arrange(p.adjust) %>% slice(1:10)
  }else if(nrow(enrich_res_gomf) < 10 ){
    gomf_plot <- enrich_res_gomf %>% arrange(p.adjust)
  }
  
  gomf_plot$Description <- gsub("_", " ", gomf_plot$Description)
  gomf_plot$Description <- factor(gomf_plot$Description, levels = rev(gomf_plot$Description))
  gomf_plot$significance <- -log10(gomf_plot$p.adjust)
  res_num <- nrow(gomf_plot)
  
  gomf_plot.p <- ggplot(gomf_plot, aes(x = significance, y = Description, color = round(p.adjust, 2))) +
      geom_point(size = 3) +
      scale_color_gradient(low = "red", high = "blue") +
      labs(x = "-log10(p.adj)", y = "", color = "p.adj", size = "Gene count", title = paste("Top ", res_num, " enriched GO-MF terms")) +
      theme_classic() +
      theme(
        axis.text.y = element_text(size = 11),
        legend.position = "right")
  ggsave(gomf_plot.p, filename = gsub("txt","gomf_plot.png",output_file),dpi = 300)
}

# KEGG
if (nrow(gene_df) > 0) {
  enrich_res_kegg <- enrichKEGG(gene = gene_df$ENTREZID,
                              organism = "hsa",
                              keyType = "ncbi-geneid",
                              pAdjustMethod = "BH",
                              minGSSize = 1,
                              pvalueCutoff = 1) %>%  
    as.data.frame() %>% mutate(Source  = "KEGG")
}else{
  enrich_res_kegg <- data.frame()
}

if(nrow(enrich_res_kegg) > 0 ){
  if(nrow(enrich_res_kegg)>= 10 ){
  kegg_plot <- enrich_res_kegg %>% arrange(p.adjust) %>% slice(1:10)
  }else if(nrow(enrich_res_kegg) < 10 ){
    kegg_plot <- enrich_res_kegg %>% arrange(p.adjust)
  }
  
  kegg_plot$Description <- gsub("_", " ", kegg_plot$Description)
  kegg_plot$Description <- factor(kegg_plot$Description, levels = rev(gomf_plot$Description))
  kegg_plot$significance <- -log10(kegg_plot$p.adjust)
  res_num <- nrow(kegg_plot)
  
  kegg_plot.p <- ggplot(kegg_plot, aes(x = significance, y = Description, color = round(p.adjust, 2))) +
      geom_point(size = 3) +
      scale_color_gradient(low = "red", high = "blue") +
      labs(x = "-log10(p.adj)", y = "", color = "p.adj", size = "Gene count", title = paste("Top ", res_num, " enriched GO-CC terms")) +
      theme_classic() +
      theme(
        axis.text.y = element_text(size = 11),
        legend.position = "right")
  ggsave(kegg_plot.p, filename = gsub("txt","kegg_plot.png",output_file),dpi = 300)
}

enrich.all <- rbind(enrich_res_gocc, enrich_res_gobp, enrich_res_gomf, enrich_res_kegg)
write.table(enrich.all, gsub("txt","enrichment.txt", output_file),quote = F,sep = "\t",row.names = F)

```

