{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee0f7acd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/_k/5d2dv9652rg7rz_wjbvxwkww0000gp/T/ipykernel_37099/4277102274.py:118: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  eqtl_catalog = pd.read_csv(cfg[\"eqtl_catalog_file\"], sep=\"\\t\", header=0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traits matching will run\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#!/usr/bin/env python3\n",
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Thu Sep 21 15:41:09 2023\n",
    "\n",
    "@author: xiyas\n",
    "\"\"\"\n",
    "\n",
    "# GenomeReportingPipeline version 3.2 on SZA sercer\n",
    "# June 2023: Adding the male/female distinguishment for \"Hemizygous\"\n",
    "# Jan 2024: vep gz file compatible\n",
    "# April 2024: Traits report, if jText[colNames_CSQ.index('Existing_variation')] in trait_list: saveFlag5 =\"Trait_var\"\n",
    "#   rs17822931&CM062373&COSV62323034 should be separated as rs17822931\n",
    "\n",
    "# version: 20240828\n",
    "# new sysmed server version 20241209: adding genebe and clingen variants\n",
    "\n",
    "# version 20241209: \n",
    "# Adding genebe and clingen variants\n",
    "\n",
    "# version: 20250214\n",
    "# The speeds, the hard filter of benign vairants\n",
    "\n",
    "#%%Cell 1 Define the path for parameters and databases\n",
    "# ==============================================================================\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import gzip\n",
    "import paramiko\n",
    "import re\n",
    "from scp import SCPClient\n",
    "import time\n",
    "from itertools import islice\n",
    "from datetime import datetime\n",
    "import pickle\n",
    "from collections import defaultdict\n",
    "\n",
    "all_start_time = time.time()\n",
    "# ====== 1) setting up paths and running parameters ======\n",
    "path = 'local'  # local / uppmax / SZA / sysmed\n",
    "\n",
    "run_traits = True\n",
    "\n",
    "# User input filtration parameters\n",
    "user_af = 0.05\n",
    "user_ada_score = 0.6\n",
    "user_rf_score = 0.6\n",
    "user_revel_score=0.75\n",
    "user_spliceai_al=0.5\n",
    "user_spliceai_dg=0.5\n",
    "user_spliceai_dl=0.5\n",
    "user_spliceai_ag=0.5\n",
    "user_bayesdel_addaf_score=0.0692655\n",
    "user_bayesdel_noaf_score=-0.0570105\n",
    "user_am_classification=['likely_pathogenic','ambigous']\n",
    "user_am_pathogenicity=0.564\n",
    "user_gender= \"Male\"\n",
    "user_clinvar = [\"Pathogenic\",\"Likely_pathogenic\",\"Uncertain_significance\",\"Benign\",\"Likely_benign\",\"Conflicting_classifications_of_pathogenicity\",\"Conflicting_interpretations_of_pathogenicity\"]\n",
    "user_acmg_classification = [\"Pathogenic\",\"Likely_pathogenic\",\"Uncertain_significance\",\"Benign\",\"Likely_benign\"]\n",
    "\n",
    "\n",
    "# config file of setting paths\n",
    "# shoudld make a new Pharma_df! ________________________________--\n",
    "config = {\n",
    "    \"local\": {\n",
    "        \"fileName\": \"/Users/xinmengliao/Documents/Project/20250516_Webserver/Case_Study/Healthy/WBWG_01_P001_30.gz_vep_annotated.vcf.gz\",\n",
    "        \"geneBaseFile\": \"/Users/xiyas/ATPM_Project/database-v3/GeneDB_GenCC.txt\",\n",
    "        #\"diseaesDBfile\": \"/Users/xiyas/V2_Genome_reporting/database-file-v2/diseaseDB_1115_3.txt\",\n",
    "        \"OMIM_inheritance_DBfile\": \"/Users/xinmengliao/Documents/Project/20250516_Webserver/Datasets/pheno_OMIM_all.txt\",\n",
    "        \"clingen_file\": \"/Users/xiyas/ATPM_Project/database-v3/Clingen-variants-2024-12-09.txt\",\n",
    "        \"Pharma_dbfile\": \"/Users/xinmengliao/Documents/Project/20250516_Webserver/Datasets/PharmGKB/clinical_annotation_combined.txt\",\n",
    "        \"Trait_dbfile\": \"/Users/xinmengliao/Documents/Project/20250516_Webserver/Datasets/Reports_genome_databases_traits_merged_2.txt\",\n",
    "        \"clinvar_vcf_path\": \"/Users/xinmengliao/Documents/Project/20250516_Webserver/Datasets/ClinVar/clinvar_20250504.vcf.gz\",\n",
    "        \"GenCC_path\" : \"/Users/xinmengliao/Documents/Project/20250516_Webserver/Datasets/GenCC/GenCCforGene.txt\",\n",
    "        \"eqtl_catalog_file\" : \"/Users/xinmengliao/Documents/Project/20250516_Webserver/Datasets/xQTL/eQTL_Catalogue/permuted_results/eqtl.catalogue.permuted.known.position.txt\",\n",
    "        \"eqtl_gtex_file\" : \"/Users/xinmengliao/Documents/Project/20250516_Webserver/Datasets/xQTL/GTEx_Analysis_v10_eQTL_updated/GTEx.all.tissue.known.gene.txt\",\n",
    "        \"output_file\": \"/Users/xinmengliao/Documents/Project/20250516_Webserver/Case_Study/Healthy/WBWG_01_P001_30.nodup.txt\"\n",
    "    },\n",
    "    \"sysmed\": {\n",
    "        \"fileName\": None,\n",
    "        \"outFile\": None,\n",
    "        \"clingen_file\": \"/mnt/storage_pool/Genomics/Genome/database-files/Clingen-variants-2024-12-09.txt\",\n",
    "        \"geneBaseFile\": \"/mnt/storage_pool/Genomics/Genome/database-files/GeneDB_GenCC.txt\",\n",
    "        \"diseaesDBfile\": \"/mnt/storage_pool/Genomics/Genome/database-files/diseaseDB_1115_3.txt\",\n",
    "        \"OMIM_inheritance_DBfile\": \"/mnt/storage_pool/Genomics/Genome/database-files/pheno_OMIM_all.txt\",\n",
    "        \"ontology_file\": \"/mnt/storage_pool/Genomics/Genome/database-files/genedb.ontology.all0307.csv\",\n",
    "        \"GWAS_dbfile\": \"/mnt/storage_pool/Genomics/Genome/database-files/Merged_GWAS_vcf_2024.txt\",\n",
    "        \"Pharma_dbfile\": \"/mnt/storage_pool/Genomics/Genome/database-files/Merged_Pharma_vcf_2024.txt\",  # 更新路径\n",
    "        \"Pharma_dbfile_all\": \"/mnt/storage_pool/Genomics/Genome/database-files/Pharma_db_all.txt\",  # 添加 Pharma_dbfile_all\n",
    "        \"Haplotype_rs_ref\": \"/mnt/storage_pool/Genomics/Genome/database-files/PGx_annotation_all.levels.drugs.hap250220.txt\",  # 更新路径\n",
    "        \"Trait_dbfile\": \"/mnt/storage_pool/Genomics/Genome/database-files/Reports_genome_databases_traits_merged_2.txt\",\n",
    "        \"clinvar_vcf_path\": \"/mnt/storage_pool/Genomics/Genome/database-files/clinvar_20240611.vcf.gz\"\n",
    "    }\n",
    "}\n",
    "\n",
    "if path not in config:\n",
    "    raise ValueError(f\"unknown path: {path}\")\n",
    "\n",
    "cfg = config[path]\n",
    "\n",
    "\n",
    "# ====== 3) Reading necessary files ======\n",
    "\n",
    "def read_db_file(filepath, encoding=\"ISO-8859-1\", sep=\"\\t\", fillna_str=\"No info\", drop_allna_cols=False):\n",
    "    \"\"\"help to formally read files\"\"\"\n",
    "    df = pd.read_csv(filepath, sep=sep, encoding=encoding)\n",
    "    if drop_allna_cols:\n",
    "        df = df.dropna(axis=1, how='all')\n",
    "    df = df.replace(np.nan, fillna_str)\n",
    "    return df\n",
    "\n",
    "OMIM_Inheritance_DB = read_db_file(cfg[\"OMIM_inheritance_DBfile\"])\n",
    "OMIM_Inheritance_DB['phenotypeMimNumber'] = OMIM_Inheritance_DB['phenotypeMimNumber'].astype(str).str.strip().str.replace(r'\\.0$', '', regex=True)\n",
    "OMIM_Inheritance_DB['inheritances'] = OMIM_Inheritance_DB['inheritances'].replace(np.nan,\"Inheritance Not provided by OMIM\")\n",
    "geneBaseFile = cfg[\"geneBaseFile\"]\n",
    "eqtl_catalog = pd.read_csv(cfg[\"eqtl_catalog_file\"], sep=\"\\t\", header=0) \n",
    "eqtl_catalog.columns = [f\"eQTL_Catalog_{col}\" for col in eqtl_catalog.columns]\n",
    "eqtl_gtex = pd.read_csv(cfg[\"eqtl_gtex_file\"], sep=\"\\t\", header=0)\n",
    "eqtl_gtex.columns = [f\"eQTL_GTEx_{col}\" for col in eqtl_gtex.columns]\n",
    "\n",
    "# ====== 4) Decision of whether running GWAS/Pharmaco/Trait ======\n",
    "if run_traits:\n",
    "    Trait_db = read_db_file(cfg[\"Trait_dbfile\"], drop_allna_cols=True)\n",
    "    trait_list = Trait_db['variants'].to_list()\n",
    "    print(\"Traits matching will run\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "49fd2276",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%% ClinVar DB dynamic updates: parse into clinvar_cache\n",
    "def parse_clinvar_vcf(clinvar_vcf_path):\n",
    "    print(f\"Parsing ClinVar VCF from {clinvar_vcf_path}...\")\n",
    "    clinvar_dict = {}      \n",
    "    with gzip.open(clinvar_vcf_path, 'rt') as file:\n",
    "        for line in file:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue  \n",
    "            cols = line.strip().split(\"\\t\")\n",
    "            chrom, pos, id_, ref, alt, qual, filter_, info = cols[:8]\n",
    "            # get needed clinvar info\n",
    "            clinvar_info = {}\n",
    "            for key in [\"CLNSIG\", \"CLNDN\", \"CLNHGVS\", \"CLNSIGINCL\", \"CLNVC\", \"GENEINFO\",\"CLNSIGCONF\", \"CLNREVSTAT\", \"CLNDNINCL\"]:\n",
    "                match = re.search(fr\"{key}=([^;]+)\", info)\n",
    "                if match:\n",
    "                    value = match.group(1)\n",
    "                    ## note this is very important that all csq single value should separate not by |\n",
    "                    # if isinstance(value, list):\n",
    "                    #     value = \"&\".join(value)  # 如果 ClinVar 解析出来是 list，则连接\n",
    "                    # **CLNDISDB 里的 `,` 也要替换成 `&`**\n",
    "                    if key == \"CLNDISDB\" or key == \"CLNDN\":\n",
    "                        value = value.replace(\",\", \"&\").replace(\"|\", \"&\")\n",
    "                else:\n",
    "                    value = None\n",
    "                clinvar_info[key] = value\n",
    "            # build the key\n",
    "            key = (chrom, pos, ref, alt)\n",
    "            clinvar_dict[key] = clinvar_info\n",
    "    return clinvar_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ec139e17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000000 lines processed!\n",
      "2000000 lines processed!\n",
      "3000000 lines processed!\n",
      "4000000 lines processed!\n",
      "Cell 2 VEP annotated File processing done! Now start to map GeneDB and DiseaseDB\n",
      "Total processing time: 40.55 seconds\n"
     ]
    }
   ],
   "source": [
    "#%%Cell 2 start the analysis program for the vep annotated vcf files, generated report A B C and D || E for the traits\n",
    "# ==============================================================================\n",
    "start_time = time.time()\n",
    "# col_map is used to build a mapping once colNames_CSQ is obtained\n",
    "col_map = {}  # Will be populated after parsing the \"ID=CSQ\" line\n",
    "headings = []\n",
    "\n",
    "## For REVEL scores refinement from 0.460&0.460&.&. to 0.460\n",
    "def extract_decimal_from_string(s):\n",
    "    if not s or not isinstance(s, str):\n",
    "        return None\n",
    "    matches = re.findall(r\"\\d+\\.\\d+\", s)\n",
    "    if not matches:\n",
    "        return None\n",
    "    # 返回第一个匹配到的数值\n",
    "    return matches[0]\n",
    "\n",
    "#####check it's male or female, to get the correct Zygosity for the chrX variants\n",
    "#check_male_flag = 0>1\n",
    "#check_male_flag = False\n",
    "\n",
    "# Reading vcf.gz file \n",
    "file = gzip.open(fileName,'rt')\n",
    "tLine = file.readline()\n",
    "#tLine = tLine.decode(\"ISO-8859-1\")\n",
    "i = 0\n",
    "reportA,reportE = [], []\n",
    "\n",
    "while tLine:\n",
    "    # remove the newline character\n",
    "    tLine = tLine.rstrip('\\n')\n",
    "    # split the current line\n",
    "    iContent = tLine.split('\\t')\n",
    "    i += 1\n",
    "    ##get the content from VCF annotation header\n",
    "    if tLine.startswith('#'):\n",
    "        if 'ID=CSQ' in tLine:\n",
    "            annoText = iContent[0].split('Format: ')\n",
    "            colNames_CSQ = annoText[1].replace('\">','')\n",
    "            colNames_CSQ = colNames_CSQ.split('|')\n",
    "            # construct col_map for all use\n",
    "            col_map = { name: idx for idx, name in enumerate(colNames_CSQ) }\n",
    "        elif tLine.startswith('#CHROM'):\n",
    "            headings = iContent\n",
    "        # directly goes into next line\n",
    "        tLine = file.readline()\n",
    "        #print(tLine)\n",
    "        continue\n",
    "    #start processing real data rows\n",
    "    if not headings:\n",
    "        tLine = file.readline()\n",
    "        #print(tLine)\n",
    "        continue\n",
    "    \n",
    "    # iContent is a VCF line => [CHROM, POS, ID, REF, ALT, QUAL, FILTER, INFO, FORMAT, sampleGenotype...]\n",
    "    # Get multiple transcripts annotation (iText)\n",
    "    # this is going to analyse all transcript result inside CSQ and any one satisfied the criteria, the whole variant goes into report A/B.\n",
    "    iText = [s for s in iContent[headings.index('INFO')].split(';') if 'CSQ=' in s]\n",
    "    iText = iText[0].replace('CSQ=','').split(',')\n",
    "    \n",
    "    \n",
    "    #### =================ClinVar version updates ##### =================\n",
    "    saveFlag1, saveFlag2,saveFlag3,saveFlag4, saveFlag5 = False, False, False, False, False\n",
    "    \n",
    "    ## 1.26 2024: In the conference I checked rs6025, which should be SZAvar360812 a famous PGx genes for F5 gene. However it only appears on nodup4files, as a clinVar genes,\n",
    "    ## not in PGx reports. so I realized this elif here should be all changed to \"if\", because elif ignores other cases if the first one satisfied the criteria\n",
    "    for j in range(0,len(iText)):\n",
    "        jText = iText[j].split('|')\n",
    "        # fixing REVEL score \n",
    "        if 'REVEL' in col_map and 'REVEL_score' in col_map:\n",
    "            revel_idx = col_map['REVEL']\n",
    "            revel_score_idx = col_map['REVEL_score']\n",
    "            if jText[revel_idx] == '':\n",
    "                parsed_value = extract_decimal_from_string(jText[revel_score_idx])\n",
    "                if parsed_value is not None:\n",
    "                    jText[revel_idx] = parsed_value\n",
    "                    #print(parsed_value)\n",
    "                    #print(jText[col_map['REVEL']])\n",
    "                            \n",
    "        # 1) ClinVar and prediction filtered based on users' needs \n",
    "        if 'MAX_AF' in col_map:\n",
    "            max_af_val = jText[col_map['MAX_AF']]\n",
    "            try:\n",
    "                max_af_numeric = float(max_af_val) if max_af_val != '' else 0.0\n",
    "            except (ValueError, TypeError):\n",
    "                max_af_numeric = 0.0\n",
    "            jText[col_map['MAX_AF']] = max_af_numeric\n",
    "        \n",
    "        if max_af_numeric < user_af:\n",
    "            \n",
    "            if any(clinvar_term in jText[col_map['ClinVar_CLNSIG']] for clinvar_term in user_clinvar if 'ClinVar_CLNSIG' in col_map):\n",
    "                saveFlag1 = \"Keep\"\n",
    "\n",
    "            predicted_impact = (\n",
    "                    ('IMPACT' in col_map and jText[col_map['IMPACT']] == 'HIGH')\n",
    "                    or ('ada_score' in col_map and jText[col_map['ada_score']] != '' and float(jText[col_map['ada_score']]) > user_ada_score)\n",
    "                    or ('rf_score' in col_map and jText[col_map['rf_score']] != '' and float(jText[col_map['rf_score']]) > user_rf_score)\n",
    "                    or ('REVEL' in col_map and jText[col_map['REVEL']] != '' and float(jText[col_map['REVEL']]) > user_revel_score)\n",
    "                    or ('SpliceAI_pred_DS_AL' in col_map and jText[col_map['SpliceAI_pred_DS_AL']] != '' and float(jText[col_map['SpliceAI_pred_DS_AL']])>user_spliceai_al)\n",
    "                    or ('SpliceAI_pred_DS_DG' in col_map and jText[col_map['SpliceAI_pred_DS_DG']] != '' and float(jText[col_map['SpliceAI_pred_DS_DG']])>user_spliceai_dg)\n",
    "                    or ('SpliceAI_pred_DS_DL' in col_map and jText[col_map['SpliceAI_pred_DS_DL']] != '' and float(jText[col_map['SpliceAI_pred_DS_DL']])>user_spliceai_dl)\n",
    "                    or ('SpliceAI_pred_DS_AG' in col_map and jText[col_map['SpliceAI_pred_DS_AG']] != '' and float(jText[col_map['SpliceAI_pred_DS_AG']])>user_spliceai_ag)\n",
    "                    or ('BayesDel_addAF_score' in col_map and jText[col_map['BayesDel_addAF_score']] != '' and float(jText[col_map['BayesDel_addAF_score']])>user_bayesdel_addaf_score)\n",
    "                    or ('BayesDel_noAF_score' in col_map and jText[col_map['BayesDel_noAF_score']] != '' and float(jText[col_map['BayesDel_noAF_score']])>user_bayesdel_noaf_score)\n",
    "                    or ('am_class' in col_map and jText[col_map['am_class']] in user_am_classification\n",
    "                        and 'am_pathogenicity' in col_map\n",
    "                        and jText[col_map['am_pathogenicity']] != ''\n",
    "                        and float(jText[col_map['am_pathogenicity']])>user_am_pathogenicity))\n",
    "            if predicted_impact:\n",
    "                saveFlag1 = \"Keep\"\n",
    "        \n",
    "        # 3) judge Trait, need run_traits=True and trait_list\n",
    "        if run_traits and 'Existing_variation' in col_map:\n",
    "            ex_variation_val = jText[col_map['Existing_variation']]\n",
    "            ex_variants = ex_variation_val.split('&')\n",
    "            if ex_variants and ex_variants[0] in trait_list:\n",
    "                saveFlag5 = \"Trait_var\"\n",
    "    # after for j in range(len(iText)) loop, if saveFlag1/2/3/4/5 has value, append the line to respective report\n",
    "    if saveFlag1:\n",
    "        reportA.append(tLine)\n",
    "    if saveFlag5:\n",
    "        reportE.append(tLine)\n",
    "\n",
    "    # print progress every 1000000 lines\n",
    "    if i % 1000000 == 0:\n",
    "        print(f\"{i} lines processed!\")\n",
    "\n",
    "    # read the next line\n",
    "    tLine = file.readline()\n",
    "\n",
    "\n",
    "file.close()\n",
    "print('Cell 2 VEP annotated File processing done! Now start to map GeneDB and DiseaseDB')\n",
    "end_time = time.time()\n",
    "print(\"Total processing time: {:.2f} seconds\".format(end_time - start_time))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1ee45b36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== 数据处理统计 ===\n",
      "原始VCF处理行数: 250\n",
      "reportA变异数: 251\n",
      "reportA展开行数: 251\n",
      "CSQ注释字段数: 102\n",
      "总输出列数: 112\n"
     ]
    }
   ],
   "source": [
    "base_vcf_columns = ['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT','Sample_Info']\n",
    "\n",
    "# CSQ注释列名（从col_map中获取，按照索引顺序排列）\n",
    "csq_columns = [''] * len(col_map)  # 初始化空列表\n",
    "for col_name, col_index in col_map.items():\n",
    "    csq_columns[col_index] = col_name\n",
    "\n",
    "# 合并所有列名：基础VCF列 + CSQ注释列\n",
    "all_output_columns = base_vcf_columns + csq_columns\n",
    "\n",
    "def process_vcf_line_to_expanded_format(vcf_line, col_map, csq_columns):\n",
    "    \"\"\"\n",
    "    将VCF行转换为展开格式，每个转录本注释作为单独一行\n",
    "    \"\"\"\n",
    "    expanded_rows = []\n",
    "\n",
    "    fields = vcf_line.split('\\t')\n",
    "    info_field = fields[7]  # INFO\n",
    "    csq_info = [s for s in info_field.split(';') if 'CSQ=' in s]\n",
    "    \n",
    "    if not csq_info:\n",
    "        return expanded_rows\n",
    "    \n",
    "    csq_data = csq_info[0].replace('CSQ=', '').split(',')\n",
    "    \n",
    "    for transcript_annotation in csq_data:\n",
    "        transcript_fields = transcript_annotation.split('|')\n",
    "        \n",
    "        # 确保transcript_fields长度与csq_columns匹配\n",
    "        while len(transcript_fields) < len(csq_columns):\n",
    "            transcript_fields.append('')\n",
    "        \n",
    "        expanded_row = fields + transcript_fields[:len(csq_columns)]\n",
    "        expanded_rows.append(expanded_row)\n",
    "    \n",
    "    return expanded_rows\n",
    "\n",
    "expanded_reportA = []\n",
    "for i, vcf_line in enumerate(reportA):\n",
    "    expanded_rows = process_vcf_line_to_expanded_format(vcf_line, col_map, csq_columns)\n",
    "    expanded_reportA.extend(expanded_rows)\n",
    "if expanded_reportA:\n",
    "    expanded_reportA = pd.DataFrame(expanded_reportA, columns=all_output_columns)\n",
    "\n",
    "\n",
    "\n",
    "# ==============================================================================\n",
    "# 数据质量检查和统计\n",
    "# ==============================================================================\n",
    "print(\"\\n=== 数据处理统计 ===\")\n",
    "print(f\"原始VCF处理行数: {i}\")\n",
    "print(f\"reportA变异数: {len(reportA)}\")\n",
    "print(f\"reportA展开行数: {len(expanded_reportA)}\")\n",
    "\n",
    "print(f\"CSQ注释字段数: {len(csq_columns)}\")\n",
    "print(f\"总输出列数: {len(all_output_columns)}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "36940b93",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:You are not logged in to GeneBe. We recommend using username and api_key arguments. Find out more on https://genebe.net/api .\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: .netrc file not found.\n",
      "Warning: Invalid format in [chr6-170561966-AGCAGCAGCAGCAGCAGCAGCAG-A,ACAGCAGCAGCAGCAGCAGCAG], not annotating.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00,  1.62it/s]\n"
     ]
    }
   ],
   "source": [
    "#%% Cell 10 GeneBe ACMG classification ======================================\n",
    "import genebe as gnb\n",
    "\n",
    "small_df = expanded_reportA.loc[:,[\"#CHROM\",\"POS\",\"REF\",\"ALT\"]]\n",
    "small_df = small_df.rename(columns={\"#CHROM\":\"chr\",\"POS\":\"pos\",\"REF\":\"ref\",\"ALT\":\"alt\"})\n",
    "unique_small_df = small_df.drop_duplicates()\n",
    "\n",
    "try:\n",
    "    annotated_df = gnb.annotate(\n",
    "        unique_small_df,\n",
    "        genome='hg38',\n",
    "        use_ensembl=False,\n",
    "        use_refseq=True,\n",
    "        flatten_consequences=True,\n",
    "        output_format=\"dataframe\"\n",
    "    )\n",
    "except Exception as e:\n",
    "    print(f\"GeneBe annotation failed: {str(e)}\")\n",
    "    # create an empty DataFrame to keep the structure\n",
    "    annotated_df = pd.DataFrame(columns=[\n",
    "        'chr', 'pos', 'ref', 'alt', 'gene_symbol', \n",
    "        'acmg_score', 'acmg_classification', 'acmg_criteria'\n",
    "    ])\n",
    "annotated_df = annotated_df.rename(columns={\"chr\":\"#CHROM\",\"pos\":\"POS\",\"ref\":\"REF\",\"alt\":\"ALT\"})\n",
    "\n",
    "required_columns = [\"#CHROM\",\"POS\",\"REF\",\"ALT\",\"gene_symbol\",\n",
    "                   \"acmg_score\",\"acmg_classification\",'acmg_criteria']\n",
    "for col in required_columns:\n",
    "    if col not in annotated_df.columns:\n",
    "        annotated_df[col] = None  # add missing columns\n",
    "\n",
    "small_annotate_all = annotated_df[required_columns].rename(columns={'gene_symbol': 'SYMBOL'})\n",
    "# merge automatically handles missing values\n",
    "expanded_reportA = pd.merge(\n",
    "    expanded_reportA, \n",
    "    small_annotate_all, \n",
    "    how=\"left\", \n",
    "    on=[\"#CHROM\",\"POS\",\"REF\",\"ALT\",\"SYMBOL\"]\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e42769bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ClinVar and ACMG scoring system\n",
    "expanded_reportA = expanded_reportA[expanded_reportA['acmg_classification'].isin(user_acmg_classification)]\n",
    "\n",
    "acmg_conditions = [\n",
    "    expanded_reportA['acmg_classification'] == 'Pathogenic',\n",
    "    expanded_reportA['acmg_classification'] == 'Likely_pathogenic',\n",
    "    expanded_reportA['acmg_classification'] == 'Benign',\n",
    "    expanded_reportA['acmg_classification'] == 'Likely_benign',\n",
    "    expanded_reportA['acmg_classification'] == 'Uncertain_significance'\n",
    "]\n",
    "acmg_choices = [1, 0.8, 0.2, 0.2, 0.5]\n",
    "\n",
    "clinvar_conditions = [\n",
    "    expanded_reportA['ClinVar_CLNSIG'].str.contains('Pathogenic', na=False, case=True),\n",
    "    expanded_reportA['ClinVar_CLNSIG'].str.contains('Likely_pathogenic', na=False, case=True),\n",
    "    expanded_reportA['ClinVar_CLNSIG'].str.contains('Benign', na=False, case=True), \n",
    "    expanded_reportA['ClinVar_CLNSIG'].str.contains('Likely_benign', na=False, case=True),\n",
    "    expanded_reportA['ClinVar_CLNSIG'].str.contains('Uncertain_significance', na=False, case=True),\n",
    "    expanded_reportA['ClinVar_CLNSIG'].str.contains('Conflicting_classifications_of_pathogenicity', na=False, case=True),\n",
    "    expanded_reportA['ClinVar_CLNSIG'].str.contains('Conflicting_interpretations_of_pathogenicity', na=False, case=True)\n",
    "]\n",
    "clinvar_choices = [1, 0.8, 0.2, 0.2, 0.5, 0.5, 0.5]\n",
    "\n",
    "expanded_reportA['ACMG_score'] = np.select(acmg_conditions, acmg_choices, default=0)\n",
    "expanded_reportA['ClinVar_score'] = np.select(clinvar_conditions, clinvar_choices, default=0)\n",
    "expanded_reportA['Pathogenicity_score'] = expanded_reportA['ACMG_score'] * expanded_reportA['ClinVar_score']\n",
    "\n",
    "# AF Score\n",
    "af_values = pd.to_numeric(expanded_reportA['MAX_AF'], errors='coerce')\n",
    "min_af = af_values[af_values > 0].min()  # 确保大于0，因为要计算log\n",
    "af_filled = af_values.fillna(min_af)\n",
    "af_filled = af_filled.replace(0, min_af)\n",
    "af_score_raw = -np.log10(af_filled)\n",
    "min_score = af_score_raw.min()\n",
    "max_score = af_score_raw.max()\n",
    "\n",
    "if max_score > min_score: \n",
    "    expanded_reportA['AF_score'] = (af_score_raw - min_score) / (max_score - min_score)\n",
    "else:\n",
    "    expanded_reportA['AF_score']\n",
    "\n",
    "# Prediction Score\n",
    "spliceai_cols = ['SpliceAI_pred_DS_AG', 'SpliceAI_pred_DS_AL', 'SpliceAI_pred_DS_DG', 'SpliceAI_pred_DS_DL']\n",
    "for col in spliceai_cols:\n",
    "    expanded_reportA[col] = pd.to_numeric(expanded_reportA[col], errors='coerce').fillna(0)\n",
    "expanded_reportA['spliceai_max'] = expanded_reportA[spliceai_cols].max(axis=1)\n",
    "\n",
    "prediction_conditions = [\n",
    "    expanded_reportA['Consequence'].str.startswith('splice_', na=False),\n",
    "    (expanded_reportA['IMPACT'] == 'HIGH') & (~expanded_reportA['Consequence'].str.startswith('splice_', na=False)),\n",
    "    expanded_reportA['Consequence'].str.startswith('missense_variant', na=False),\n",
    "    (expanded_reportA['IMPACT'] != 'LOW') & (~expanded_reportA['Consequence'].str.startswith('missense_', na=False))\n",
    "]\n",
    "\n",
    "expanded_reportA['am_pathogenicity'] = pd.to_numeric(expanded_reportA['am_pathogenicity'], errors='coerce').fillna(0)\n",
    "prediction_choices = [expanded_reportA['spliceai_max'], 0.99, 0.66 * expanded_reportA['am_pathogenicity'], 0.33]\n",
    "expanded_reportA['Prediction_score'] = np.select(prediction_conditions, prediction_choices, default=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f2c145bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# eQTL\n",
    "expanded_reportA['variant_info'] = expanded_reportA[['#CHROM', 'POS', 'REF', 'ALT']].fillna('').astype(str).agg('_'.join, axis=1)\n",
    "\n",
    "def eqtl_score_fun(data: pd.DataFrame, eqtl_catalog: pd.DataFrame, eqtl_gtex: pd.DataFrame) -> pd.DataFrame:\n",
    "    \n",
    "    final_data = data.merge(\n",
    "        eqtl_catalog, \n",
    "        left_on='variant_info', \n",
    "        right_on='eQTL_Catalog_variant', \n",
    "        how='left'\n",
    "    ).drop_duplicates()\n",
    "    \n",
    "    final_data = final_data.merge(\n",
    "        eqtl_gtex,\n",
    "        left_on='variant_info',\n",
    "        right_on='eQTL_GTEx_variant_id',\n",
    "        how='left'\n",
    "    ).drop_duplicates()\n",
    "    \n",
    "    return final_data\n",
    "\n",
    "expanded_reportA = eqtl_score_fun(expanded_reportA, eqtl_catalog, eqtl_gtex)\n",
    "expanded_reportA = expanded_reportA.drop_duplicates()\n",
    "\n",
    "expanded_reportA.to_csv(cfg[\"output_file\"], sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9018c7d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current Time = 13:48:46\n",
      "processing 102 lines of traits\n",
      "rs4343This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4341This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs671This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs890022This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1513475This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs12367809This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs563296This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs558129This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs9877408This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs11549465This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1800795This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs901850This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs11121542This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs6905419This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4253778This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs8192678This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs303760This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs13135092This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs56068671This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs3843540This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4626333This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4343This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4341This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs2854464This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1042713This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1042714This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs12883788This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs17602729This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs10501089This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs11548200This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs3213537This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs11881338This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4134943This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs55680124This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs558129This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs11549465This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs7247312This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1800795This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs205262This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs8011870This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs572934641This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4253778This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs2920503This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs13107325This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4880This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs10505025This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4734621This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4343This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4341This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1042713This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1042714This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4994This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs17602729This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs9939609This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs7181866This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs558129This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1695This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1799945This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs11549465This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1572312This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4253778This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs8192678This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs7732671This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs10499043This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1937This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1870377This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1800012This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs2854464This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1800169This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1042713This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4253778This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1051168This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs5400This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs16991615This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1799941This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1800795This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs8179183This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs17300539This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1042714This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs7138803This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs12970134This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs174547This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4654748This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs33972313This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs2282679This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs12272004This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs17822931This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1805007This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs12821256This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1805007This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1805008This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1805009This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs7349332This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs3827760This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs3739070This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs2003046This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1426654This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs713598This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs10034228This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1953558This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4481887This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs671This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs2472297This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs73598374This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs4988235This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs140926439This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs429358This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs7412This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs34637584This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs76763715This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs574307253This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs116956554This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs6948912This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs117713019This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs35424804This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs17688916This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs3764002This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs1444978This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs78446248This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs7570This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs117890891This patient carries wildtype genotype (keep as homo_ref)\n",
      "rs6725323This patient carries wildtype genotype (keep as homo_ref)\n",
      "Current Time = 13:48:46\n",
      "T/T\n",
      "rs429358_gt TT\n",
      "rs7412 CC\n",
      "get APOE genotype from original trait file\n",
      "Done. Check: /Users/xinmengliao/Documents/Project/20250516_Webserver/Case_Study/Healthy/WBWG_01_P001_30.nodup_traits.txt\n"
     ]
    }
   ],
   "source": [
    "#%% Cell 7 Traits report ====================================\n",
    "## note: one SNP (rsID) could match to more than one trait phenotypes. such as rs1815739. In the result file we only output one db matching but we do the multiple match in the updated_traits.csv\n",
    "if run_traits == True:\n",
    "    from datetime import datetime\n",
    "    now = datetime.now()\n",
    "    newLineHeadings = ['Traits name', 'category', 'genes', 'variants', 'Description','Patient genotypes','Genotype Description','Zygosity'];\n",
    "    newLine = '\\t'.join(newLineHeadings)+'\\n'\n",
    "    with open(cfg[\"output_file\"].replace('.txt','_traits.txt'),'w') as f:\n",
    "        f.write(newLine)\n",
    "    current_time = now.strftime(\"%H:%M:%S\")\n",
    "    print(\"Current Time =\", current_time)\n",
    "    print('processing ' + str(len(reportE)) + ' lines of traits' )\n",
    "    #this irs_id_set is used to record the trait variants this individual have, and later write the one don't have to a \"wild type\"\n",
    "    irs_id_set = {s.split('|')[col_map['Existing_variation']].split('&')[0] for s in reportE}\n",
    "    for i in range(0,len(reportE)):\n",
    "        current_time = now.strftime(\"%H:%M:%S\")\n",
    "        ##iText means all transcripts\n",
    "        iText = [s for s in reportE[i].split('\\t')[7].split(';') if 'CSQ=' in s][0].replace('CSQ=','').split(',')\n",
    "        irs_id = iText[0].split(\"|\")[col_map['Existing_variation']].split('&')[0]\n",
    "        iREF = reportE[i].split('\\t')[3].split(',')\n",
    "        iALT = reportE[i].split('\\t')[4].split(',')\n",
    "        iGenoTypeList = iREF+iALT\n",
    "        if \".\" not in reportE[i].split('\\t')[9].split(':')[0]:\n",
    "            iGenotypInd1 = int(reportE[i].split('\\t')[9][0])\n",
    "            iGenotypInd2 = int(reportE[i].split('\\t')[9][2])\n",
    "            #iGenotype = iGenoTypeList[0]+'/'+ iGenoTypeList[0] + '>' + iGenoTypeList[iGenotypInd1] +'/'+iGenoTypeList[iGenotypInd2]\n",
    "            try:\n",
    "                iGenotype = iGenoTypeList[0]+'/'+ iGenoTypeList[0] + '>' + iGenoTypeList[iGenotypInd1] +'/'+iGenoTypeList[iGenotypInd2]\n",
    "            except:\n",
    "                iGenotype = iREF[0]+'>'+iALT[0]+':'+reportE[i].split('\\t')[9][0]+reportE[i].split('\\t')[9][1]+reportE[i].split('\\t')[9][2]\n",
    "            iTemp = reportE[i].replace('\\n','').split('\\t')\n",
    "            iTemp[headings.index('INFO')] = iTemp[headings.index('INFO')].split(';CSQ=')[0]\n",
    "            #sys.exit('!')\n",
    "            if iGenotypInd1 == iGenotypInd2:\n",
    "                iZygo = 'Homozygous'\n",
    "            elif iGenotypInd1 == 0 or iGenotypInd2 == 0:\n",
    "                iZygo = 'Heterozygous'\n",
    "            else:\n",
    "                iZygo = 'Compound heterozygous'\n",
    "            ##get the Pharma and sza id from sub transcripts, named jText\n",
    "            ## here we seems not care about genes, transcripts, only focuse on variant so I deleted for loop\n",
    "            iTraits= Trait_db[Trait_db['variants']== irs_id]['Traits name'].tolist()[0]\n",
    "            iCategory = Trait_db[Trait_db['variants']== irs_id]['category'].tolist()[0]\n",
    "            iGenes= Trait_db[Trait_db['variants']== irs_id]['genes'].tolist()[0]\n",
    "            iDescription = Trait_db[Trait_db['variants']== irs_id]['Description'].tolist()[0].strip()\n",
    "            iGenoType_description =Trait_db[Trait_db['variants']== irs_id]['Genotype_Description'].tolist()[0].strip()\n",
    "            newLine_temp ='\\t'.join([iTraits,iCategory,iGenes,irs_id,iDescription,iGenotype,iGenoType_description,iZygo])+'\\n'\n",
    "            with open(cfg[\"output_file\"].replace('.txt','_traits.txt'),'a') as f:\n",
    "                f.write(newLine_temp)\n",
    "    # Iterate through Trait_db and write entries not in reportE\n",
    "    for index, row in Trait_db.iterrows():\n",
    "        if row['variants'] not in irs_id_set:\n",
    "            print(row['variants'] + \"This patient carries wildtype genotype (keep as homo_ref)\")\n",
    "            newLine_temp = '\\t'.join([row['Traits name'], row['category'], row['genes'], row['variants'], row['Description'], row['REF']+'/'+row['REF'], row['Genotype_Description'], 'homo_ref']) + '\\n'\n",
    "            with open(cfg[\"output_file\"].replace('.txt', '_traits.txt'), 'a') as f:\n",
    "                f.write(newLine_temp)\n",
    "    now = datetime.now()\n",
    "    current_time = now.strftime(\"%H:%M:%S\")\n",
    "    print(\"Current Time =\", current_time) \n",
    "\n",
    "#%% APOE haploytype based on generated results file\n",
    "def get_apoe_isoform_and_risk(rs429358_genotype, rs7412_genotype):\n",
    "    \"\"\"\n",
    "    based on rs429358_genotype and rs7412_genotype\n",
    "    return (APOE Isoform, AD risk description) tuple.\n",
    "\n",
    "    reference table:\n",
    "    rs429358(Ref:T)  rs7412(Ref:C)   Isoform   AD risk\n",
    "**  TT        TT       ε2/ε2    Reduced Risk (protective)\n",
    "?   TT        CT       ε2/ε3    Neutral to Reduced Risk\n",
    "?   CC        CT       ε2/ε3    Neutral to Reduced Risk\n",
    "**  TT        CC       ε3/ε3    Neutral (most common)  \n",
    "?   TC        CC       ε3/ε4    Increased Risk\n",
    "**  CC        CC       ε4/ε4    Highest Risk\n",
    "    \"\"\"\n",
    "    genotype_map = {\n",
    "        (\"TT\", \"TT\"): (\"e2/e2\", \"Reduced Risk (protective)\"),\n",
    "        (\"TT\", \"CT\"): (\"e2/e3\", \"Neutral to Reduced Risk\"),\n",
    "        (\"TT\", \"CC\"): (\"e3/e3\", \"Neutral (most common)\"),\n",
    "        (\"TC\", \"CC\"): (\"e3/e4\", \"Increased Risk\"),\n",
    "        (\"CC\", \"CC\"): (\"e4/e4\", \"Highest Risk\"),\n",
    "    }\n",
    "    return genotype_map.get((rs429358_genotype, rs7412_genotype), (\"Unknown\", \"Unknown\"))\n",
    "\n",
    "    \n",
    "rows = []\n",
    "trait_outfile = cfg[\"output_file\"].replace('.txt','_traits.txt')\n",
    "\n",
    "\n",
    "# 2) 打开原文件，按行读\n",
    "with open(trait_outfile, 'r', encoding='utf-8') as f_in:\n",
    "    lines = f_in.readlines()\n",
    "\n",
    "rs429358_gt = None\n",
    "rs7412_gt = None\n",
    "\n",
    "for line in lines:\n",
    "    parts = line.strip().split('\\t')\n",
    "    if len(parts) < 6:\n",
    "        continue  # skip incomplete lines\n",
    "    \n",
    "    # parts[3] => 'rs429358' or 'rs7412'\n",
    "    variant = parts[3]\n",
    "    # parts[5] => 'Patient genotypes' (e.g. \"T/T\", \"C/C\")\n",
    "    genotype_str = parts[5]\n",
    "    if '>' in genotype_str:\n",
    "        genotype = genotype_str.split('>')[1]  # get the part after '>'\n",
    "    else:\n",
    "        genotype = genotype_str  # no '>', use the original string directly\n",
    "\n",
    "    genotype = genotype.replace('/', '')  \n",
    "    if variant == 'rs429358':\n",
    "        rs429358_gt = genotype\n",
    "        print(parts[5])\n",
    "        print(\"rs429358_gt\",rs429358_gt)\n",
    "    elif variant == 'rs7412':\n",
    "        rs7412_gt = genotype\n",
    "        print(\"rs7412\",rs7412_gt)\n",
    "\n",
    "# 3) if both are got, then calculate Isoform & Risk\n",
    "if rs429358_gt and rs7412_gt:\n",
    "    print(\"get APOE genotype from original trait file\")\n",
    "    isoform, risk = get_apoe_isoform_and_risk(rs429358_gt, rs7412_gt)\n",
    "    newLine_temp_parts = [\n",
    "        \"Alzheimer's disease risk\",                     # Traits name\n",
    "        \"Neurogenic and Cognitive functions\",          # category\n",
    "        \"APOE\",                                        # genes\n",
    "        \"rs429358+rs7412\",                               # variants\n",
    "        \"Combined APOE genotype from rs429358+rs7412\", # Description\n",
    "        isoform,                  # Patient genotypes\n",
    "        risk,                                       # Genotype Description\n",
    "        \"\"                                           # Zygosity\n",
    "    ]\n",
    "    newLine_temp = '\\t'.join(newLine_temp_parts) + '\\n'\n",
    "\n",
    "    with open(trait_outfile, 'a', encoding='utf-8') as f_out:\n",
    "        f_out.write(newLine_temp)\n",
    "    print(f\"Done. Check: {trait_outfile}\")\n",
    "else:\n",
    "    isoform, risk = (\"Unknown\", \"Unknown\")\n",
    "    print(\"did not find two genptypes of rs429358 and rs7412\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
