---
title: "Single Variant Management"
author: "Xinmeng Liao"
date: "2025-06-13"
output: html_document
---

```{r setup, include=FALSE}
library(dplyr)
library(tidyverse)
library(data.table)
library(epigraphdb)
for (i in c("select","filter", "mutate","rename", "left_join", "slice")){
  conflicted::conflict_prefer(i, "dplyr")
}
rm(i)
conflicted::conflicts_prefer(stats::sd)
conflicted::conflicts_prefer(httr::content)
conflicted::conflicts_prefer(plotly::layout)
```

# Load file
```{r}
result_file <- "/Users/xinmengliao/Documents/Project/20250516_Webserver/usr_output/output_text.txt"
result <- read.csv(result_file,header = T,sep = "\t")

colnames(result) <- gsub("am_","AlphaMissense_", colnames(result))
result <- result %>% 
  mutate(rsID = sapply(strsplit(split = "&", Existing_variation),`[`,1)) %>% 
  mutate(rsID = if_else(grepl("^rs", rsID,fixed = F), rsID, NA)) %>% unique() 
output_file <- "/Users/xinmengliao/Documents/Project/20250516_Webserver/usr_output/Variant/SingleVariant/singlevariant_output.txt"

# Other files
gwas <- fread("/Users/xinmengliao/Documents/Project/20250516_Webserver/Datasets/gwas_catalog_v1.0-associations_e113_r2025-04-28.tsv", 
            header = T, sep = "\t", quote = "")
eqtl_catalogue_tissue <- read.csv("/Users/xinmengliao/Documents/Project/20250516_Webserver/Datasets/xQTL/eQTL_Catalogue/eqtl_catalogue_tissue_lables.tsv",
                                header = T, sep = "\t", quote = "")
geneid <- read.csv("/Users/xinmengliao/Documents/Project/20250516_Webserver/Datasets/xQTL/eQTL_Catalogue/ensemblID_gene.txt", 
                 header = T, sep = "\t") %>% select(1, 5) %>% unique()
trait <- read.csv("/Users/xinmengliao/Documents/Project/20250516_Webserver/Datasets/Reports_genome_databases_traits_merged_2.txt", 
            header = T, sep = "\t")
pharmgkb_data <- read.csv("/Users/xinmengliao/Documents/Project/20250516_Webserver/Datasets/PharmGKB/clinical_annotation_combined.txt", 
                        header = T, sep = "\t") 
```

# xQTL: eQTL and pQTL
```{r}
rsid <- result %>% select(Existing_variation) %>% as.character()
rsid <- unlist(strsplit(rsid, split = "&"))[grep("^rs", unlist(strsplit(rsid, split = "&")))]
    
if(length(rsid) > 0) {
  tryCatch({
    url <- paste0("https://www.ebi.ac.uk/eqtl/api/v1/associations/", rsid)
    response <- GET(url)
    if (status_code(response) == 200) {
      data_json <- content(response, as = "text", encoding = "UTF-8")
      parsed <- fromJSON(data_json)
      eqtl_catalogue_df <- rbindlist(parsed$`_embedded`$associations)
      eqtl_catalogue_df <- eqtl_catalogue_df %>% 
        left_join(., eqtl_catalogue_tissue, by = c("tissue" = "tissue_id")) %>% unique() %>% 
        left_join(., geneid, by = c("gene_id" = "Gene.stable.ID")) %>% unique()
      eqtl_catalogue_col <- c("variant", "Gene.name", "ref", "beta", "r2", "se", "pvalue", "tissue_label", "study_id", "qtl_group")
      eqtl_catalogue_df <- eqtl_catalogue_df %>% select(all_of(eqtl_catalogue_col)) %>% 
        rename(Variant = variant, Gene = Gene.name, `Ref allele` = ref, Tissue = tissue_label, Beta = beta, SE = se, R2 = r2, Pvalue = pvalue, `Study ID` = study_id, `Group` = qtl_group)
    } else {
      cols <- c('Variant', 'Gene', 'Beta', 'R2', 'SE', 'Pvalue', 'Ref.allele', 'Tissue', 'Study ID', 'eQTL group')
      eqtl_catalogue_df <- as_tibble(setNames(as.list(rep(NA, length(cols))), cols))
    }
  }, error = function(e) {
    cols <- c('Variant', 'Gene', 'Beta', 'R2', 'SE', 'Pvalue', 'Ref.allele', 'Tissue', 'Study ID', 'eQTL group')
    eqtl_catalogue_df <<- as_tibble(setNames(as.list(rep(NA, length(cols))), cols))
  })
} else {
  cols <- c('Variant', 'Gene', 'Beta', 'R2', 'SE', 'Pvalue', 'Ref.allele', 'Tissue', 'Study ID', 'eQTL group')
  eqtl_catalogue_df <- as_tibble(setNames(as.list(rep(NA, length(cols))), cols))
}



rela.pro <- pqtl_pleio(rsid = rsid)

if(nrow(rela.pro) > 0) {
  pqtl_df <- data.frame()
  for (j in 1:nrow(rela.pro)) {
    pqtl_tpm <- pqtl(query = rela.pro$expID[j], searchflag = "proteins", rtype = "mrres", pvalue = 1) 
    temp_df <- pqtl_tpm %>% 
      select(rsID, expID, outID, trans_cis, beta, se, pvalue, ld_check, method) %>% 
      rename(Variant = rsID, Protein = expID, `Disease/Phenotype` = outID, Type = trans_cis, 
             Beta = beta, SE = se, Pvalue = pvalue, `Linkage disequilibrium` = ld_check, Method = method)
    pqtl_df <- rbind(pqtl_df, temp_df)
  }
} else {
  cols <- c('Variant', 'Protein', 'Disease/Phenotype', 'Type', 'Beta', 'SE', 'Pvalue', 'Linkage disequilibrium', 'Method')
  pqtl_df <- as_tibble(setNames(as.list(rep(NA, length(cols))), cols))
  pqtl_df <- pqtl_df %>% mutate(Variant = rsid)

}

xqtl.df <- eqtl_catalogue_df %>% left_join(., pqtl_df, by = "Variant")

write.table(xqtl.df, gsub("txt","xqtl.txt", output_file),quote = F,sep = "\t",row.names = F)
```

# GWAS and traits and PGx
```{r}
# GWAS
gwas.df <- gwas %>% filter(SNPS == rsid)
      
if(nrow(gwas.df) > 0) {
  gwas.df <- gwas.df %>% 
    select(SNPS, DISEASE.TRAIT, CHR_ID, CHR_POS, RISK.ALLELE.FREQUENCY, P.VALUE, OR.or.BETA, X95..CI..TEXT., LINK) %>% 
    rename(`Disease/Traits` = DISEASE.TRAIT, Chr = CHR_ID, Pos = CHR_POS, Variant = SNPS,
           `Risk allele frequency` = RISK.ALLELE.FREQUENCY, 
           `P-value` = P.VALUE, `Odd ratio/beta` = OR.or.BETA, 
           `95% Confidence Interval` = X95..CI..TEXT., Source = LINK) 
} else {
  gwas.cols <- c("Variant","Disease/Traits", "Chr", "Pos", "Risk allele frequency", "P-value", "Odd ratio/beta", "95% Confidence Interval", "Source")
  gwas.df <- as_tibble(setNames(as.list(rep(NA, length(gwas.cols))), gwas.cols))
}


# traits
trait.df <- trait %>% filter(variants == rsid)
if(nrow(trait.df) > 0) {
  trait.df <- trait.df %>% 
    select(variants, genes, Traits.name, category, Genotype_Description, Description) %>% 
    rename(`Variant` = variants, Genes = genes, Traits = Traits.name, Category = category, 
           `Genotype description` = Genotype_Description, `Trait details` = Description)
} else {
  trait.cols <- c("Variant", "Genes", "Traits", "Category", "Genotype description", "Trait details")
  trait.df <- as_tibble(setNames(as.list(rep(NA, length(trait.cols))), trait.cols))
}


if(length(rsid) > 0) {
  pharmgkb_df <- pharmgkb_data %>% 
    filter(Variant.Haplotypes == rsid) %>% 
    select(-Latest.History.Date..YYYY.MM.DD., -Level.Override, -Score, -PMID.Count, -Evidence.Count) %>% 
    select(Variant.Haplotypes, Gene, Drug.s., Phenotype.s., Level.of.Evidence, Level.Modifiers, 
           Genotype.Allele, Annotation.Text, Allele.Function, Specialty.Population)
} else {
  cols <- c('Variant.Haplotypes', 'Gene', 'Drug.s.', 'Phenotype.s.', 'Level.of.Evidence',
            'Level.Modifiers', 'Genotype.Allele', 'Annotation.Text', 'Allele.Function', 'Specialty.Population')
  pharmgkb_df <- as_tibble(setNames(as.list(rep(NA, length(cols))), cols))
}

trait.all <- gwas.df %>% left_join(., trait.df, by = "Variant") %>% 
  left_join(., pharmgkb_df, by = c("Variant" = "Variant.Haplotypes"))
write.table(trait.all, gsub("txt","gwas_trait_pgx.txt", output_file),quote = F,sep = "\t",row.names = F)

```

